### PR TITLE
Implement Brave Publisher Ads

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/7298

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

See https://github.com/brave/brave-browser/issues/7298 and we can arrange a call where needed to demonstrate this new feature

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: browser/extensions/api/brave_rewards_api.cc [hunk 1]
```diff
@@ -608,6 +608,9 @@ ExtensionFunction::ResponseAction BraveRewardsSaveAdsSettingFunction::Run() {
       const auto is_enabled =
           params->value == "true" && ads_service_->IsSupportedLocale();
       ads_service_->SetEnabled(is_enabled);
+    } else if (params->key == "shouldShowPublisherAdsOnParticipatingSites") {
```
Comment (masparrow) [inline browser/extensions/api/brave_rewards_api.cc]: Can you use `kShouldShowPublisherAdsOnParticipatingSites`?

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 2]
```diff
@@ -1335,6 +1343,8 @@ void RewardsDOMHandler::SaveAdsSetting(const base::ListValue* args) {
     const auto is_enabled =
         value == "true" && ads_service_->IsSupportedLocale();
     ads_service_->SetEnabled(is_enabled);
+  } else if (key == kShouldShowPublisherAdsOnParticipatingSites) {
+    ads_service_->SetShowPublisherAdsOnPariticipatingSites(value == "true");
```
Comment (masparrow) [inline browser/ui/webui/brave_rewards_page_ui.cc]: Can we have a `kTrue` (or better named) const string for these?

File: browser/extensions/api/brave_rewards_api.cc (hunk 1 above)
Comment (tmancey) [reply browser/extensions/api/brave_rewards_api.cc]: Fixed

File: browser/ui/webui/brave_rewards_page_ui.cc (hunk 2 above)
Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Fixed

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 3]
```diff
@@ -305,9 +305,14 @@ class RewardsDOMHandler : public WebUIMessageHandler,
...
 namespace {
 
+const char kShouldShowPublisherAdsOnParticipatingSites[] =
+    "shouldShowPublisherAdsOnParticipatingSites";
+
 const char kShouldAllowAdConversionTracking[] =
     "shouldAllowAdConversionTracking";
 
+const char kTrue[] = "true";
+
```
Comment (masparrow) [inline browser/ui/webui/brave_rewards_page_ui.cc]: nit: Ah now, you know what I'm going to say... defining `kTrue` in multiple places... tsk... :D can we avoid it easily for this PR?

File: components/services/bat_ads/public/cpp/ads_client_mojo_bridge.cc [hunk 4]
```diff
@@ -58,6 +58,22 @@ void AdsClientMojoBridge::IsEnabled(IsEnabledCallback callback) {
   std::move(callback).Run(ads_client_->IsEnabled());
 }
 
+bool AdsClientMojoBridge::ShouldShowPublisherAdsOnPariticipatingSites(
+    bool* is_enabled) {
+  if (!is_enabled) {
```
Comment (masparrow) [inline components/services/bat_ads/public/cpp/ads_client_mojo_bridge.cc]: nit: while valid - it doesn't read 'valid' ... perhaps just DCHECK `is_enabled` here?

Comment (masparrow) [reply components/services/bat_ads/public/cpp/ads_client_mojo_bridge.cc]: or explicit `is_enabled == nullptr`

Comment (tmancey) [reply components/services/bat_ads/public/cpp/ads_client_mojo_bridge.cc]: Changed to `DCHECK` to match Google Style Guide

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 5]
```diff
@@ -1405,6 +1524,99 @@ CreativeAdNotificationList AdsImpl::GetAdsForUnseenAdvertisers(
   return unseen_ads;
 }
 
+CreativePublisherAdList AdsImpl::GetEligiblePublisherAds(
```
Comment (masparrow) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: nit: Given `GetUnseenPublisherAdsAndRoundRobinIfNeeded` below, should this be

`GetEligiblePublisherAdsAndRoundRobinIfNeeded`

?

File: vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc [hunk 6]
```diff
@@ -41,10 +41,12 @@ Result CatalogState::FromJson(
   new_catalog_id = catalog["catalogId"].GetString();
 
   new_version = catalog["version"].GetUint64();
-  if (new_version != 1) {
-    // TODO(Terry Mancey): Implement Log (#44)
-    // 'patch invalid', { reason: 'unsupported version', version: version }
-    return SUCCESS;
+  if (new_version != kCatalogVersion) {
+    if (error_description != nullptr) {
+      *error_description = "Unsupported catalog verion";
```
Comment (masparrow) [inline vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc]: nit: typo - `verion`

File: browser/ui/webui/brave_rewards_page_ui.cc (hunk 3 above)
Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Not at this time as out of scope of this PR, but we do need to look at sharing these values

File: vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc (hunk 6 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 5 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: Refactored but these names are correct

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 7]
```diff
@@ -17,6 +17,7 @@
 #include "base/task/thread_pool/thread_pool_instance.h"
 #include "bat/ads/issuers_info.h"
 #include "bat/ads/ad_notification_info.h"
+#include "bat/ads/publisher_ad_info.h"
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: where do we use this include?

Comment (tmancey) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: Removed

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 8]
```diff
@@ -553,6 +565,100 @@ AdsHistory AdsImpl::GetAdsHistory(
...
+    ad.creative_instance_id = creative_publisher_ad.creative_instance_id;
+    ad.creative_set_id = creative_publisher_ad.creative_set_id;
+    ad.category = creative_publisher_ad.category;
+    ad.size = creative_publisher_ad.size;
+    ad.creative_url = creative_publisher_ad.creative_url;
+    ad.target_url = creative_publisher_ad.target_url;
+    ads.entries.push_back(ad);
+
+    BLOG(INFO) << "Publisher ad for URL: " << url
+        << std::endl << "  creativeInstanceId: " << ad.creative_instance_id
```
Comment (gdregalo) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: `std::endl` flushes the buffer every time.  Use `\n` instead.

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: @gdregalo: Legacy so need to change in many places, will raise a new issue for this

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 9]
```diff
@@ -174,6 +175,17 @@ - (void)setEnabled:(BOOL)enabled
   }
 }
 
+- (BOOL)shouldShowPublisherAdsOnParticipatingSites
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm]: This property matches the method `shouldShowPublisherAdsOnParticipatingSites` from the AdsClient so we'll have to change one of them or else this won't compile (dupe symbol)

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 10]
```diff
@@ -396,6 +415,16 @@ - (void)getCreativeAdNotifications:(const std::vector<std::string> &)categories
   callback(ads::Result::SUCCESS, categories, found_ads);
 }
 
+- (void)getCreativePublisherAds:(const std::string &)url categories:(const std::vector<std::string> &)categories sizes:(const std::vector<std::string> &)sizes callback:(ads::OnGetCreativePublisherAdsCallback)callback
+{
+  // TODO(brave): To be implemented
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm]: What needs to be implemented here?

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 11]
```diff
@@ -396,6 +415,16 @@ - (void)getCreativeAdNotifications:(const std::vector<std::string> &)categories
...
+}
+
+- (void)isParticipatingSiteForPublisherAds:(const std::string &)url callback:(ads::OnIsParticipatingSiteForPublisherAdsCallback)callback
+{
+  // TODO(brave): To be implemented
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm]: What needs to be implemented here?

File: vendor/brave-ios/Ads/BATBraveAds.mm (hunk 10 above)
Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm]: Updated comment. However, this work is not required until Publisher Ads are required on iOS

File: vendor/brave-ios/Ads/BATBraveAds.mm (hunk 11 above)
Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm]: Updated comment. However, this work is not required until Publisher Ads are required on iOS

File: vendor/brave-ios/Ads/BATBraveAds.mm (hunk 9 above)
Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm]: Fixed

File: browser/extensions/api/brave_rewards_api.cc [hunk 12]
```diff
@@ -34,6 +34,15 @@ using brave_rewards::RewardsServiceFactory;
 namespace extensions {
 namespace api {
 
+namespace {
```
Comment (NejcZdovc) [inline browser/extensions/api/brave_rewards_api.cc]: You do not need to change anything here as it will not be used in the panel

Comment (tmancey) [reply browser/extensions/api/brave_rewards_api.cc]: now that I have had to rebase again I will test publisher ads are working as expected then remove the UI and put PR up for review, thanks

**Comment by tmancey** [comment]
Closing as stale

**Pull request closed without merging**